src/core/thread/meshcop_dataset_manager.cpp

```diff
@@ -0,0 +1,544 @@
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements common methods for manipulating MeshCoP Datasets.
+ *
+ */
+
+#include <stdio.h>
+
+#include <coap/coap_header.hpp>
+#include <common/code_utils.hpp>
+#include <common/logging.hpp>
+#include <common/timer.hpp>
+#include <platform/random.h>
+#include <thread/meshcop_dataset.hpp>
+#include <thread/meshcop_dataset_manager.hpp>
+#include <thread/thread_netif.hpp>
+#include <thread/thread_tlvs.hpp>
+#include <thread/thread_uris.hpp>
+
+namespace Thread {
+namespace MeshCoP {
+
+DatasetManager::DatasetManager(ThreadNetif &aThreadNetif, const char *aUri):
+    mMle(aThreadNetif.GetMle()),
+    mNetif(aThreadNetif),
+    mNetworkDataLeader(aThreadNetif.GetNetworkDataLeader()),
+    mResource(aUri, &HandleSet, this),
+    mTimer(&HandleTimer, this),
+    mUri(aUri),
+    mCoapServer(aThreadNetif.GetCoapServer())
+{
+    mCoapServer.AddResource(mResource);
+}
+
+ThreadError DatasetManager::Set(const Dataset &aDataset, uint8_t &aFlags)
+{
+    ThreadError error = kThreadError_None;
+
+    aFlags = 0;
+
+    VerifyOrExit(mNetwork.GetTimestamp().Compare(aDataset.GetTimestamp()) > 0, error = kThreadError_InvalidArgs);
+
+    mLocal = aDataset;
+    aFlags |= kFlagLocalUpdated;
+
+    switch (mMle.GetDeviceState())
+    {
+    case Mle::kDeviceStateChild:
+    case Mle::kDeviceStateRouter:
+        if (mNetwork.GetTimestamp().Compare(mLocal.GetTimestamp()) > 0)
```

Comment:
This checks whether the local Active Operational Dataset is newer than the version that the network is advertising.  If newer, then start a timer to register the local version with the Leader.